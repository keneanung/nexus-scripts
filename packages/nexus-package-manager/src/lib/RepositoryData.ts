export type RepositoryData = PackageEntry[];
export interface PackageEntry {
  name: string;
  packageName: string;
  description: string;
  url: string;
  dependencies: string[];
  website?: string;
  version: string;
}
