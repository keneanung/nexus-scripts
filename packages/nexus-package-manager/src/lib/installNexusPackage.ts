export const installNexusPackage = (pkgJson: any) => {
  const newPkg = globalThis.nexusclient.packages().create(pkgJson.name, pkgJson.description);
  const reflexes = globalThis.nexusclient.reflexes();
  newPkg.apply(pkgJson, reflexes);
  reflexes.run_function('onInstall', undefined, pkgJson.name);
};
