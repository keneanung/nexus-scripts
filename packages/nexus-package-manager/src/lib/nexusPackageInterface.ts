export const installNexusPackage = (pkgJson: any) => {
  const newPkg = globalThis.nexusclient.packages().create(pkgJson.name, pkgJson.description);
  const reflexes = globalThis.nexusclient.reflexes();
  newPkg.apply(pkgJson, reflexes);
  reflexes.run_function('onInstall', undefined, pkgJson.name);
};

export const isNexusPackageInstalled = (packageName: string) => {
  return globalThis.nexusclient.packages().exists(packageName, true) !== 0;
};

export const uninstallNexusPackage = (packageName: string) => {
  globalThis.nexusclient.reflexes().run_function('onUninstall', undefined, packageName);
  globalThis.nexusclient.packages().remove(packageName);
};
