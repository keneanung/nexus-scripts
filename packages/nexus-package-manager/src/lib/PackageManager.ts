import { installNexusPackage } from "./installNexusPackage";
import { RepositoryData } from "./RepositoryData";

export class PackageManager{

  private repositoryData: RepositoryData = [];

  public updateAsync = async () => {
    const response = await fetch('https://keneanung.github.io/nexus-package-repository/repository.json');
    this.repositoryData = await response.json();
  }

  public getRepositoryData = () => this.repositoryData;
  public installAsync = async (packageName: string) => {
    const pkgEntry = this.repositoryData.find(entry => entry.packageName === packageName);
    if(!pkgEntry){
      throw new Error("");

    }
    const response = await fetch(pkgEntry.url);
    const pkgJson = await response.json();

    installNexusPackage(pkgJson);
  }
}
