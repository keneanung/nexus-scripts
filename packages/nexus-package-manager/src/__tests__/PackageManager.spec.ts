import { PackageManager } from '../index';
import { enableFetchMocks } from 'jest-fetch-mock'
import { installNexusPackage } from '../lib/installNexusPackage';
enableFetchMocks()
jest.mock('../lib/installNexusPackage');
const installNexusPackageMock = jest.mocked(installNexusPackage);

beforeEach(() => {
  // cleaning up the mess left behind the previous test
  fetchMock.resetMocks();
  installNexusPackageMock.mockReset();
});

test("Should request the correct URL", async () => {
  fetchMock.mockResponse('[]');
  const sut = new PackageManager();

  await sut.updateAsync();

  expect(fetchMock).toHaveBeenCalledWith('https://keneanung.github.io/nexus-package-repository/repository.json');
});

test('Should parse response correctly', async () => {
  fetchMock.mockResponse('[{"name":"Event Bus","packageName":"eventbus","description":"Event Bus similar to the Nexus function system, but available outside of the GUI","url":"https://keneanung.github.io/nexus-event-bus/EventBus.nxs","dependencies":[]}]')
  const sut = new PackageManager();

  await sut.updateAsync();

  expect(sut.getRepositoryData()).toMatchSnapshot();
});

test('Should update correctly if called consecutely', async () => {
  fetchMock.mockResponses(
    '[{"name":"first response","packageName":"first","description":"This is the first response","url":"https://keneanung.github.io/nexus-event-bus/first.nxs","dependencies":["firstFoo"]}]',
    '[{"name":"second response","packageName":"second","description":"This is the second response","url":"https://keneanung.github.io/nexus-event-bus/second.nxs","dependencies":["secondFoo"]}]'
  );
  const sut = new PackageManager();

  await sut.updateAsync();
  await sut.updateAsync();

  expect(sut.getRepositoryData()).toMatchSnapshot();
});

test('Should have installed package after call to install', async() => {
  fetchMock.mockResponseOnce('[{"name":"second response","packageName":"second","description":"This is the second response","url":"https://keneanung.github.io/nexus-event-bus/foo.nxs","dependencies":["secondFoo"]}]');
  fetchMock.mockResponseOnce('{"name": "foo", "description": "bar"}');
  const sut = new PackageManager();
  await sut.updateAsync();

  await sut.installAsync('second');

  expect(installNexusPackageMock).toMatchSnapshot();
})