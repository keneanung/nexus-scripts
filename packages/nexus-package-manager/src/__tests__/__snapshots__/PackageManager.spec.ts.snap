// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ask the Nexus client to uninstall and then install a new version of a package with a given name on update of package 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "description": "bar",
        "name": "foo",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should have installed package after call to install 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "description": "bar",
        "name": "foo",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should parse response correctly 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "description": "Event Bus similar to the Nexus function system, but available outside of the GUI",
    "name": "Event Bus",
    "packageName": "eventbus",
    "url": "https://keneanung.github.io/nexus-event-bus/EventBus.nxs",
  },
]
`;

exports[`Should update correctly if called consecutely 1`] = `
Array [
  Object {
    "dependencies": Array [
      "secondFoo",
    ],
    "description": "This is the second response",
    "name": "second response",
    "packageName": "second",
    "url": "https://keneanung.github.io/nexus-event-bus/second.nxs",
  },
]
`;
